feat: Added Access Advisor for IAM

amanenk opened the pull request:
🎉 Thank you for making CloudQuery awesome by submitting a PR 🎉

#### Summary

closes cloudquery/cloudquery#1389 

---

Use the following steps to ensure your PR is ready to be reviewed

- [x] Read the [contribution guidelines](../blob/main/CONTRIBUTING.md) 🧑‍🎓
- [x] Run `go fmt` to format your code 🖊
- [x] Lint your changes via `golangci-lint run --new-from-rev main` 🚨 (install golangci-lint [here](https://golangci-lint.run/usage/install/#local-installation))
- [x] Update or add tests. Learn more about testing [here](https://docs.cloudquery.io/docs/developers/sdk/testing) 🧪
- [x] Update the docs by running `go run ./docs/docs.go` and committing the changes 📃
- [x] If adding a new resource, add [relevant Terraform files](../terraform) in a separate PR 📂
- [x] Ensure the status checks below are successful ✅

amanenk committed:
feat: Added Access Advisor for IAM groups

amanenk committed:
migrations added

bbernays commented on resources/services/iam/groups_access_advisor.go:
```diff
@@ -0,0 +1,188 @@
+		DeleteFilter:  client.DeleteAccountFilter,
+		IgnoreInTests: true,
+		Options:       schema.TableCreationOptions{PrimaryKeys: []string{"account_id", "service_namespace"}},
+		Columns: []schema.Column{
```
Will the table structure differ for IAM Roles/Policies/Users or could we make the table generic and just have specific resolvers for each type? I think that would make it easier for querying/using the data

amanenk replied:
You are right. The structure is the same, but before listing access advisor resources we need to list users, groups, etc. So root fetch should be different. Some of the resolvers can be generic. I have added a commit that does it. Thanks for a hint.

amanenk committed:
generic resolvers

amanenk committed:
fix

amanenk committed:
linter

amanenk committed:
Merge remote-tracking branch 'upstream/main' into access_advisor

amanenk committed:
linter

cq-bot committed:
chore: Synced file(s) with cloudquery/.github (#934)

Synced local file(s) with [cloudquery/.github](https://github.com/cloudquery/.github).





---

This PR was created automatically by the [repo-file-sync-action](https://github.com/BetaHuhn/repo-file-sync-action) workflow run [#2371502887](https://github.com/cloudquery/.github/actions/runs/2371502887)

bbernays commented on resources/services/iam/groups_access_advisor.go:
```diff
@@ -0,0 +1,178 @@
+		IgnoreError:   client.IgnoreAccessDeniedServiceDisabled,
+		DeleteFilter:  client.DeleteAccountFilter,
+		IgnoreInTests: true,
+		Options:       schema.TableCreationOptions{PrimaryKeys: []string{"account_id", "service_namespace"}},
```
If we are going to use the same table for multiple entities then the PKs need something to differentiate between groups/roles and polices

bbernays replied:
But if all of the data is the same except for a single column at the root table don't you think it would be simpler from a user perspective when querying?

bbernays replied:
I can see tradeoffs on both sides of this. So I want to get others opinions

@roneli @zagronitay @yevgenypats  

We are working on implementing [4 resources](https://github.com/cloudquery/cq-provider-aws/issues/730) that have exactly the same fields. Do you think it is better to put the data into a single table and have a field in the root table to identify which resource type it is or have 4 different tables and infer the resource type by the table name?

amanenk committed:
added access advisor for users, roles, policies

amanenk committed:
docs, gen fix

bbernays commented on resources/provider/migrations/postgres/32_v0.11.8.down.sql:
```diff
@@ -0,0 +1,7 @@
+-- CHANGEME: Verify or edit this file before proceeding
+
+-- Resource: iam.access_advisor_details
+DROP TABLE IF EXISTS aws_iam_access_advisor_tracked_actions_last_accessed;
```
I think the migrations can be removed now

bbernays commented on resources/services/iam/groups_access_advisor_details.go:
```diff
@@ -0,0 +1,184 @@
+		Resolver:      fetchIamGroupsAccessAdvisorDetails,
+		Multiplex:     client.AccountMultiplex,
+		IgnoreError:   client.IgnoreAccessDeniedServiceDisabled,
+		DeleteFilter:  AccessAdvisorFilter("group"),
```
```suggestion
		DeleteFilter:  AccessAdvisorFilter(GROUP),
```

bbernays commented on resources/services/iam/roles_access_advisor_details.go:
```diff
@@ -0,0 +1,184 @@
+		IgnoreInTests: true,
+		Options:       schema.TableCreationOptions{PrimaryKeys: []string{"account_id", "service_namespace"}},
+		Columns: []schema.Column{
+			{
+				Name:        "account_id",
+				Description: "The AWS Account ID of the resource.",
```
Rather than duplicating all of the column definitions in each resource can we create a single place for them and then just reference it? This way it will be easier for people to understand that it is a single table and if they make changes in one file it will impact multiple resources

bbernays reviewed (commented):
Left a few comments, but overall when I run it on a clean install I get the following error:
```
Type: Internal Severity: Error
        Summary: table name aws_iam_access_advisor_detail_tracked_actions_last_accessed used more than once, duplicates are in iam.users_access_advisor_details and iam.roles_access_advisor_details
```

disq commented:
The duplicate name error on `aws_iam_access_advisor_detail_tracked_actions_last_accessed` comes from the fact that we're using the same table for multiple resources as results. It needs to be a separate name or an unrelational table that's defined only once. Having the same name isn't really handled anywhere so unless we overhaul the SDK and core to allow that, overriding the test wouldn't work as it'll just blow up somewhere else unexpectedly.

yevgenypats commented:
This needs removal of migrations and update to latest SDK

yevgenypats commented:
Should go to the monorepo https://github.com/cloudquery/cq-provider-aws/pull/922 if we still want to pursue it.
